File: dash/backend/app/schemas/scim.py
"""Administrator-facing SCIM token, mapping, and log schemas."""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, Field, model_validator

from app.models.enums import UserRole


class ScimTokenCreate(BaseModel):
    name: str = Field(min_length=1, max_length=255)
    expires_in_days: int | None = Field(default=None, ge=1, le=3650)


class ScimTokenRead(BaseModel):
    id: uuid.UUID
    name: str
    token_prefix: str
    has_secret: bool = True
    created_at: datetime
    expires_at: datetime
    revoked_at: datetime | None
    last_used_at: datetime | None
    last_used_ip: str | None


class ScimTokenIssued(ScimTokenRead):
    token: str


class ScimGroupMappingUpdate(BaseModel):
    role: UserRole | None = None
    grants_all_sites: bool = False
    site_ids: list[uuid.UUID] = Field(default_factory=list, max_length=500)
    asset_group_ids: list[uuid.UUID] = Field(default_factory=list, max_length=500)

    @model_validator(mode="after")
    def validate_site_mode(self) -> ScimGroupMappingUpdate:
        if self.grants_all_sites and self.site_ids:
            raise ValueError("site_ids must be empty when grants_all_sites is true")
        if len(set(self.site_ids)) != len(self.site_ids):
            raise ValueError("site_ids must not contain duplicates")
        if len(set(self.asset_group_ids)) != len(self.asset_group_ids):
            raise ValueError("asset_group_ids must not contain duplicates")
        return self


class ScimGroupMappingRead(BaseModel):
    id: uuid.UUID
    external_id: str | None
    display_name: str
    member_count: int
    role: UserRole | None
    grants_all_sites: bool
    site_ids: list[uuid.UUID]
    asset_group_ids: list[uuid.UUID]
    created_at: datetime
    updated_at: datetime


class ScimMappingPreview(BaseModel):
    group_id: uuid.UUID
    affected_users: int
    role: UserRole | None
    grants_all_sites: bool
    site_ids: list[uuid.UUID]
    asset_group_ids: list[uuid.UUID]
    users: list[dict[str, object]]


class ScimProvisioningLogRead(BaseModel):
    id: uuid.UUID
    operation: str
    resource_type: str | None
    resource_id: str | None
    external_id: str | None
    status_code: int
    succeeded: bool
    detail: str | None
    request_id: str | None
    source_ip: str | None
    changes: dict[str, object]
    created_at: datetime


class ScimLogPage(BaseModel):
    items: list[ScimProvisioningLogRead]
    total: int
    limit: int
    offset: int

File: dash/frontend/src/types/scim.ts
import type { Role } from './auth';

export interface ScimToken {
  id: string;
  name: string;
  token_prefix: string;
  has_secret: boolean;
  created_at: string;
  expires_at: string;
  revoked_at: string | null;
  last_used_at: string | null;
  last_used_ip: string | null;
}

export interface ScimTokenIssued extends ScimToken {
  token: string;
}

export interface ScimGroupMapping {
  id: string;
  external_id: string | null;
  display_name: string;
  member_count: number;
  role: Role | null;
  grants_all_sites: boolean;
  site_ids: string[];
  created_at: string;
  updated_at: string;
}

export interface ScimMappingPayload {
  role: Role | null;
  grants_all_sites: boolean;
  site_ids: string[];
}

export interface ScimMappingPreview extends ScimMappingPayload {
  group_id: string;
  affected_users: number;
  users: Array<{
    id: string;
    email: string;
    role: Role;
    site_access_mode: 'all' | 'assigned';
  }>;
}

export interface ScimProvisioningLog {
  id: string;
  operation: string;
  resource_type: string | null;
  resource_id: string | null;
  external_id: string | null;
  status_code: number;
  succeeded: boolean;
  detail: string | null;
  request_id: string | null;
  source_ip: string | null;
  changes: Record<string, unknown>;
  created_at: string;
}

export interface ScimLogPage {
  items: ScimProvisioningLog[];
  total: number;
  limit: number;
  offset: number;
}
